crates/node/engine/src/attributes.rs

```diff
@@ -0,0 +1,324 @@
+    }
+
+    /// Checks that withdrawals for a block and attributes match.
+    pub fn check_withdrawals(
+        config: &RollupConfig,
+        attributes: &OpAttributesWithParent,
+        block: &Block<Transaction>,
+    ) -> Self {
+        let attr_withdrawals = attributes.attributes.payload_attributes.withdrawals.as_ref();
+        let attr_withdrawals = attr_withdrawals.map(|w| Withdrawals::new(w.to_vec()));
+        let block_withdrawals = block.withdrawals.as_ref();
+
+        if config.is_canyon_active(block.header.timestamp) {
+            // In canyon, the withdrawals list should be some and empty
+            if attr_withdrawals.is_none_or(|w| !w.is_empty()) {
+                return Self::Mismatch(AttributesMismatch::CanyonWithdrawalsNotEmpty);
+            }
+            if block_withdrawals.is_none_or(|w| !w.is_empty()) {
+                return Self::Mismatch(AttributesMismatch::CanyonWithdrawalsNotEmpty);
+            }
+            if !config.is_isthmus_active(block.header.timestamp) {
+                // In canyon, the withdrawals root should be set to the empty value
+                let empty_hash = alloy_consensus::EMPTY_ROOT_HASH;
+                if block.header.inner.withdrawals_root != Some(empty_hash) {
+                    return Self::Mismatch(AttributesMismatch::CanyonNotEmptyHash);
+                }
+            }
+        } else {
+            // In bedrock, the withdrawals list should be None
+            if attr_withdrawals.is_some() {
+                return Self::Mismatch(AttributesMismatch::BedrockWithdrawals);
+            }
+        }
+
+        if config.is_isthmus_active(block.header.timestamp) {
+            // In isthmus, the withdrawals root must be set
+            if block.header.inner.withdrawals_root.is_none() {
+                return Self::Mismatch(AttributesMismatch::IsthmusMissingWithdrawalsRoot);
+            }
+        }
+
+        Self::Match
+    }
+
+    /// Checks if the specified [`OpAttributesWithParent`] matches the specified [`Block`].
+    /// Returns [`AttributesMatch::Match`] if they match, otherwise returns
+    /// [`AttributesMatch::Mismatch`].
+    pub fn check(
+        config: &RollupConfig,
+        attributes: &OpAttributesWithParent,
+        block: &Block<Transaction>,
+    ) -> Self {
+        if attributes.parent.block_info.hash != block.header.inner.parent_hash {
+            return AttributesMismatch::ParentHash(
+                attributes.parent.block_info.hash,
+                block.header.inner.parent_hash,
+            )
+            .into();
+        }
+
+        if attributes.attributes.payload_attributes.timestamp != block.header.inner.timestamp {
+            return AttributesMismatch::Timestamp(
+                attributes.attributes.payload_attributes.timestamp,
+                block.header.inner.timestamp,
+            )
+            .into();
+        }
+
+        let difficulty: B256 = block.header.inner.difficulty.into();
```

Comment:
This should be `mix_hash`; `difficulty` is always `0`'d out after the merge. Can validate w/ a `cast block --json | jq`.